assemble/Particle_Diagnostics.F90

```diff
@@ -0,0 +1,1558 @@
+    enddo
+
+  end subroutine local_to_global
+
+  subroutine multi_delete_particles(mult, node_part_count, node_particles, group_arrays, summed_particles, cap_percent)
+    !Subroutine to call delete particles multiple times based on the mult factor
+    !calculated from the current number of particles and the maximum threshold
+    !> Factor to determine number of delete_particle calls
+    integer, intent(in) :: mult
+    !> Linked list of particles which exist on this node
+    type(detector_linked_list), intent(inout), dimension(:) :: node_particles
+    !> Array counting particles on all nodes
+    real, intent(inout) :: node_part_count
+    !> Number of particle arrays present
+    integer, intent(in), dimension(:) :: group_arrays
+    !> Array to sum deleted particles
+    integer, dimension(:), intent(inout) :: summed_particles
+    !> Parameter to determine if spawning/deleting will be capped per group 
+    real, optional, intent(in) :: cap_percent
+
+    integer :: i, power, j
+    integer, allocatable, dimension(:) :: remove_particles
+    logical :: power_set
+
+    allocate(remove_particles(size(group_arrays)))
+
+    power_set=.false.
+    j=0
+    do while (power_set.eqv..false.)
+       if (mult>=2**j) then
+          j=j+1
+       else
+          power=j-1
+          power_set=.true.
+       end if
+    end do
+
+    do i = 1,power
+       call delete_particles(node_part_count, node_particles, group_arrays, remove_particles, cap_percent)
+       summed_particles=summed_particles-remove_particles
+    end do
+
+    deallocate(remove_particles)
+
+  end subroutine multi_delete_particles
+
+  subroutine delete_particles(node_part_count, node_particles, group_arrays, remove_particles, cap_percent)
+    !Subroutine to delete particles in a control volume based off the number of particles present
+
+    use particles, only: particle_lists
+    !> Linked list of particles which exist on this node
+    type(detector_linked_list), intent(inout), dimension(:) :: node_particles
+    !> Array counting particles on all nodes
+    real, intent(inout) :: node_part_count
+    !> Number of particle arrays present
+    integer, intent(in), dimension(:) :: group_arrays
+    !> Array to sum number of deleted particles per particle group
+    integer, dimension(:), intent(inout) :: remove_particles
+    !> Parameter to determine if spawning/deleting will be capped per group
+    real, optional, intent(in) :: cap_percent
+
+    type(detector_type), pointer :: particle
+    type(detector_type), pointer :: temp_part
+
+    logical :: delete_group
+    integer :: j
+    real :: random
+
+    !Check ratio of each particle group present and which groups will be deleted
+    delete_group = .true.
+    if (present(cap_percent)) then
+       delete_group = .false.
+       do j=1,size(group_arrays)
+          if ((((node_particles(j)%length*1.0)/sum(node_particles(:)%length)*1.0)*100)>=cap_percent) then
```

Comment:
I should probably restrict deleting to this group so it's consistent with the spawning case. In reality if we have cap_percent on and the condition is met, only particles from this group should exist on the node anyway.

Reply:
I've decided to leave this as is. In the spawning case, if a group reaches the cap (e.g. 99%), we are saying we exist in that group's domain and only want to spawn particles from said group. Conversely in the deleting case, we are again saying we are in the 99% group's domain, and it would be beneficial to remove spurious particles from the other group if they exist here. Hence we delete across all groups instead of just the dominant group. Additionally, if we only deleted the dominant group here, we could be artificially reducing the ratio of the dominant group to other groups.

Reply:
Okay, this makes sense!